File: android/app/src/main/kotlin/xyz/gloryolaifa/stitch/engine/EngineHost.kt
package xyz.gloryolaifa.stitch.engine

import android.content.Context
import android.util.Log
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.view.TextureRegistry
import java.util.UUID
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Implements the Pigeon host API: owns the preview player and running
 * exports, and reports back through [EngineFlutterApi] on the main thread.
 * Mirrors EngineHost.swift.
 */
class EngineHost(
  private val context: Context,
  messenger: BinaryMessenger,
  private val textures: TextureRegistry,
) : EngineHostApi {
  private val callbacks = EngineFlutterApi(messenger)
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
  private var preview: PreviewPlayer? = null
  private var document: EngineDocument? = null
  private val jobs = mutableMapOf<String, Pair<EngineJob, Exporter.Listener>>()

  private fun player(): PreviewPlayer = preview ?: PreviewPlayer(context, textures) { state ->
    scope.launch { runCatching { callbacks.onPlaybackState(state) } }
  }.also { preview = it }

  override fun createPreview(): Long = player().textureId

  override fun setDocument(json: String) {
    val doc = try {
      EngineDocument.decode(json)
    } catch (e: Exception) {
      throw FlutterError("bad_document", e.message)
    }
    document = doc
    player().setDocument(doc)
  }

  override fun play() = player().play()

  override fun pause() = player().pause()

  override fun seek(positionUs: Long, exact: Boolean) = player().seek(positionUs, exact)

  override fun release() {
    preview?.release()
  }

  override suspend fun probe(path: String): MediaInfoMessage = bridged { MediaProbe.probe(path) }

  override suspend fun thumbnails(
    path: String,
    timesUs: List<Long>,
    maxSize: Long,
    outDir: String,
  ): List<String?> = Thumbnailer.thumbnails(path, timesUs, maxSize.toInt(), outDir)

  override suspend fun createProxy(path: String, outPath: String) =
    bridged { ProxyMaker.createProxy(context, path, outPath) }

  override fun capabilities(): CapabilitiesMessage = Exporter.capabilities()

  override suspend fun waveform(path: String, peaksPerSecond: Long): List<Double> =
    bridged { Waveform.peaks(path, peaksPerSecond.toInt()) }

  override fun setPreviewVolume(volume: Double) {
    preview?.setVolume(volume)
  }

  override fun startExport(request: ExportRequestMessage): String {
    val doc = document ?: throw FlutterError("bad_document", "No document to export")
    ExportService.start(context, request.progressTitle)
    return start(Exporter(context, doc, request), foreground = true)
  }

  override fun startSpeechAudio(documentJson: String, outputPath: String): String {
    val doc = try {
      EngineDocument.decode(documentJson)
    } catch (e: Exception) {
      throw FlutterError("bad_document", e.message)
    }
    return start(SpeechAudio(context, doc, outputPath))
  }

  /**
   * Runs [job], reporting through the export callbacks under a new id.
   * With [foreground], the export service shows its progress and stops
   * with it.
   */
  private fun start(job: EngineJob, foreground: Boolean = false): String {
    val jobId = UUID.randomUUID().toString()
    val listener = object : Exporter.Listener {
      override fun onProgress(fraction: Double) {
        if (foreground) ExportService.update(context, fraction)
        scope.launch { runCatching { callbacks.onExportProgress(jobId, fraction) } }
      }

      override fun onCompleted(path: String) {
        jobs.remove(jobId)
        if (foreground) ExportService.stop(context)
        scope.launch { runCatching { callbacks.onExportCompleted(jobId, path) } }
      }

      override fun onFailed(error: EngineException) {
        jobs.remove(jobId)
        if (foreground) ExportService.stop(context)
        scope.launch {
          runCatching { callbacks.onExportFailed(jobId, error.code, error.message ?: "") }
        }
      }
    }
    jobs[jobId] = job to listener
    // Start after returning the id, so Dart knows the job before any event.
    scope.launch {
      try {
        job.start(listener)
      } catch (e: Exception) {
        // Anything unexpected ends this job, not the app.
        Log.e(TAG, "Job failed to start", e)
        listener.onFailed(EngineException.exportFailed(e.message ?: "Could not start"))
      }
    }
    return jobId
  }

  override fun cancelExport(jobId: String) {
    val (job, listener) = jobs[jobId] ?: return
    job.cancel(listener)
  }

  fun dispose() {
    jobs.values.toList().forEach { (job, listener) -> job.cancel(listener) }
    preview?.dispose()
    preview = null
    scope.cancel()
  }

  private suspend fun <T> bridged(block: suspend () -> T): T = try {
    block()
  } catch (e: EngineException) {
    throw FlutterError(e.code, e.message)
  }

  companion object {
    private const val TAG = "StitchEngine"

    fun register(context: Context, engine: FlutterEngine): EngineHost {
      val messenger = engine.dartExecutor.binaryMessenger
      val host = EngineHost(context.applicationContext, messenger, engine.renderer)
      EngineHostApi.setUp(messenger, host)
      return host
    }
  }
}

File: ios/Runner/Engine/EngineHost.swift
import Flutter
import UIKit
import Foundation

/// Implements the Pigeon host API: owns the preview player and running
/// exports, and reports back through [EngineFlutterApi] on the main thread.
final class EngineHost: EngineHostApi {
  private let textures: FlutterTextureRegistry
  private let callbacks: EngineFlutterApi
  private var preview: PreviewPlayer?
  private var document: EngineDocument?
  private var jobs: [String: EngineJob] = [:]

  init(messenger: FlutterBinaryMessenger, textures: FlutterTextureRegistry) {
    self.textures = textures
    callbacks = EngineFlutterApi(binaryMessenger: messenger)
  }

  static func register(with registrar: FlutterPluginRegistrar) {
    let host = EngineHost(messenger: registrar.messenger(), textures: registrar.textures())
    // The message handlers keep the host alive for the engine's lifetime.
    EngineHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: host)
  }

  private func player() -> PreviewPlayer {
    if let preview { return preview }
    let p = PreviewPlayer(textures: textures) { [weak self] state in
      Task { @MainActor in try? await self?.callbacks.onPlaybackState(state: state) }
    }
    preview = p
    return p
  }

  func createPreview() throws -> Int64 { player().textureId }

  func setDocument(json: String) throws {
    do {
      let doc = try EngineDocument.decode(json)
      document = doc
      player().setDocument(doc)
    } catch {
      throw PigeonError(code: "bad_document", message: "\(error)", details: nil)
    }
  }

  func play() throws { player().play() }

  func pause() throws { player().pause() }

  func seek(positionUs: Int64, exact: Bool) throws {
    player().seek(us: positionUs, exact: exact)
  }

  func release() throws { preview?.release() }

  func probe(path: String) async throws -> MediaInfoMessage {
    do { return try await MediaProbe.probe(path: path) } catch let e as EngineError {
      throw PigeonError(code: e.code, message: e.message, details: nil)
    }
  }

  func thumbnails(path: String, timesUs: [Int64], maxSize: Int64, outDir: String) async throws
    -> [String?]
  {
    await Thumbnailer.thumbnails(
      path: path, timesUs: timesUs, maxSize: Int(maxSize), outDir: outDir)
  }

  func createProxy(path: String, outPath: String) async throws {
    do { try await ProxyMaker.createProxy(path: path, outPath: outPath) } catch let e as EngineError
    {
      throw PigeonError(code: e.code, message: e.message, details: nil)
    }
  }

  func capabilities() throws -> CapabilitiesMessage { Exporter.capabilities() }

  func waveform(path: String, peaksPerSecond: Int64) async throws -> [Double] {
    do { return try await Waveform.peaks(path: path, peaksPerSecond: Int(peaksPerSecond)) } catch
      let e as EngineError
    {
      throw PigeonError(code: e.code, message: e.message, details: nil)
    }
  }

  func setPreviewVolume(volume: Double) throws { preview?.setVolume(volume) }

  func startExport(request: ExportRequestMessage) throws -> String {
    guard let document else {
      throw PigeonError(code: "bad_document", message: "No document to export", details: nil)
    }
    return start(Exporter(doc: document, request: request))
  }

  func startSpeechAudio(documentJson: String, outputPath: String) throws -> String {
    let doc: EngineDocument
    do {
      doc = try EngineDocument.decode(documentJson)
    } catch {
      throw PigeonError(code: "bad_document", message: "\(error)", details: nil)
    }
    return start(SpeechAudio(doc: doc, outputPath: outputPath))
  }

  /// Runs [job], reporting through the export callbacks under a new id.
  /// An export asks for time to finish in the background; when that runs
  /// out, it stops as interrupted.
  private func start(_ job: EngineJob) -> String {
    let jobId = UUID().uuidString
    jobs[jobId] = job
    let callbacks = self.callbacks
    let background = BackgroundTime()
    if let exporter = job as? Exporter {
      background.begin { exporter.interrupt() }
    }
    Task { @MainActor [weak self] in
      defer { background.end() }
      var lastReported = -1.0
      do {
        let path = try await job.run { fraction in
          // Report at most every 1 percent.
          guard fraction - lastReported >= 0.01 || fraction >= 1 else { return }
          lastReported = fraction
          Task { @MainActor in
            try? await callbacks.onExportProgress(jobId: jobId, fraction: fraction)
          }
        }
        try? await callbacks.onExportCompleted(jobId: jobId, outputPath: path)
      } catch let e as EngineError {
        try? await callbacks.onExportFailed(jobId: jobId, code: e.code, message: e.message)
      } catch {
        try? await callbacks.onExportFailed(
          jobId: jobId, code: "export_failed", message: "\(error)")
      }
      self?.jobs[jobId] = nil
    }
    return jobId
  }

  func cancelExport(jobId: String) throws { jobs[jobId]?.cancel() }
}

/// Time to keep working in the background. When the system is about to
/// take it away, [onExpire] runs and the time is handed back at once, as
/// iOS requires. Both calls are safe from any thread.
private final class BackgroundTime {
  private var id = UIBackgroundTaskIdentifier.invalid

  func begin(onExpire: @escaping () -> Void) {
    id = UIApplication.shared.beginBackgroundTask(withName: "Export") { [weak self] in
      onExpire()
      self?.end()
    }
  }

  func end() {
    guard id != .invalid else { return }
    UIApplication.shared.endBackgroundTask(id)
    id = .invalid
  }
}
